File: 2023/python/day13/main.py
from pathlib import Path

SAMPLE_INPUT = """#.##..##.
..#.##.#.
##......#
##......#
..#.##.#.
..##..##.
#.#.##.#.

#...##..#
#....#..#
..##..###
#####.##.
#####.##.
..##..###
#....#..#
"""
SAMPLE_INPUT_B = SAMPLE_INPUT
INPUT = Path("data.txt").read_text()


def str_mismatches(a: str, b: str) -> int:
    return sum(ac != bc for ac, bc in zip(a, b))


def find_reflection_index(note: list[str], allowed_mismatches: int = 0) -> int | None:
    for i in range(1, len(note)):
        mirrored_row_pairs = zip(note[i:], reversed(note[:i]))
        mismatches = sum(str_mismatches(a, b) for a, b in mirrored_row_pairs)
        if mismatches == allowed_mismatches:
            return i
    return None


def score(note: list[str], allowed_mismatches: int = 0) -> int:
    if mirror_row := find_reflection_index(note, allowed_mismatches):
        return mirror_row * 100
    transposed = ["".join(col) for col in zip(*note)]
    if mirror_column := find_reflection_index(transposed, allowed_mismatches):
        return mirror_column
    raise ValueError("No reflection found:\n" + "\n".join(note))


def parse_input(input: str):
    return (note.splitlines() for note in input.split("\n\n"))


def part_a(input: str):
    return sum(score(note) for note in parse_input(input))


def part_b(input: str):
    return sum(score(note, 1) for note in parse_input(input))


def test_part_a(capsys):
    assert part_a(SAMPLE_INPUT) == 405
    with capsys.disabled():
        print(f"Part A: {part_a(INPUT)}")


def test_part_b(capsys):
    assert part_b(SAMPLE_INPUT_B) == 400
    with capsys.disabled():
        print(f"Part B: {part_b(INPUT)}")

File: 2023/src/bin/day13/main.rs
use std::borrow::Cow;

use advent_2023::{BasicSolution, Solution};
use anyhow::anyhow;

struct Day;

impl BasicSolution for Day {
    const INPUT: &'static str = include_str!("data.txt");
    const SAMPLE_INPUT: &'static str = include_str!("sample.txt");

    type Shared = Vec<Vec<&'static [u8]>>;
    type Answer = usize;

    const SAMPLE_ANSWER_A: Self::TestAnswer = 405;
    const SAMPLE_ANSWER_B: Self::TestAnswer = 400;

    fn shared(input: &'static str) -> anyhow::Result<Self::Shared> {
        Ok(input
            .split("\n\n")
            .map(|note| note.lines().map(str::as_bytes).collect())
            .collect())
    }

    fn part_a(notes: Cow<Self::Shared>) -> anyhow::Result<Self::Answer> {
        notes.iter().map(|note| score(note, 0)).sum()
    }

    fn part_b(notes: Self::Shared) -> anyhow::Result<Self::Answer> {
        notes.iter().map(|note| score(note, 1)).sum()
    }
}

fn score(note: &[&[u8]], allowed_mismatches: usize) -> anyhow::Result<usize> {
    if let Some(mirror_row) = find_reflection_index(
        note.iter().map(|row| row.iter().copied()),
        allowed_mismatches,
    ) {
        Ok(mirror_row * 100)
    } else if let Some(mirror_column) = find_reflection_index(
        (0..note[0].len()).map(|i| note.iter().map(move |row| row[i])),
        allowed_mismatches,
    ) {
        Ok(mirror_column)
    } else {
        Err(anyhow!("No mirror found for note\n{note:?}"))
    }
}

fn find_reflection_index<T: Iterator<Item = u8> + Clone>(
    note: impl DoubleEndedIterator<Item = T> + ExactSizeIterator + Clone,
    allowed_mismatches: usize,
) -> Option<usize> {
    (1..note.len()).find(|&i| {
        note.clone()
            .take(i)
            .rev()
            .zip(note.clone().skip(i))
            .map(|(a, b)| a.zip(b).filter(|(a, b)| a != b).count())
            .sum::<usize>()
            == allowed_mismatches
    })
}

fn main() -> anyhow::Result<()> {
    Day::main()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a() -> anyhow::Result<()> {
        Day::test_part_a()
    }

    #[test]
    fn b() -> anyhow::Result<()> {
        Day::test_part_b()
    }
}
